### DISCUSSION TITLE
Connect from not C++ world

### DISCUSSION THREAD

**Discussion by makarchuk** [actor-framework/actor-framework#1213] [Q&A]
Sorry to bother you and feel free to close this issue if it's not a right place for it.

While working with https://github.com/zeek/zeek I decided that I want to handle workers using `go` and they use `caf` for connection of external programs (https://github.com/zeek/broker), hence I was wondering is it possible to connect to actor framework from other languages (at the moment I'm interested at go). What I need to do this is either minimal API that will handle network layer for me, or some info on wire protocol (Is there docs somewhere? Is it stable?). 

Sure, full bindings are an option as well, but `go` has it's own actor system and it wouldn't be welcomed by community to have some threads being ran by `C++` code and our own pool of goroutines next to it (and not very efficient either, I think). I feel like connecting will be more idiomatic way of doing it.

**Comment by Neverlord** [comment]
Welcome to CAF! 🙂

> feel free to close this issue if it's not a right place for it

No worries, questions are always welcome here.

> I was wondering is it possible to connect to actor framework from other languages

I'm afraid the short answer is *no*. For several reasons:
- CAF allows users to use arbitrary C++ types in messages. Reverse engineering custom C++ types in another language would only lead to very brittle interfaces.
- The protocol CAF uses internally is rather straightforward to implement. However, the reimplementation would have to simulate the *behavior* of a CAF node and mimic being part of the distributed actor system. The middle layer of CAF is fully aware of the actors using the layer. The nodes not only exchange messages, but also monitor (or spawn) remote actors for example.

Rather than trying to talk to CAF directly, it's generally a better idea to target a specific CAF application. In your case, wrapping the `Broker` API in Go. You will end up running some C++ threads internally, but I think a clean Go wrapper that hides all of that behind a channel interface would result in an idiomatic wrapper that doesn't introduce too much overhead.

Usually the best alternative for CAF applications is to provide a language-agnostic interface via REST/JSON, Protobuf, or something along those lines. This clearly separates the "internal actor world" from the "outside world" and generally gives the best interoperability. Currently, such interfaces still require a lot of manual plumbing, but we'll definitely going to tackle this topic by including more building blocks or tools for such standard interfaces.

**Comment by makarchuk** [comment]
Ok. Thank you. Will dig into API!
